### PR TITLE
Global Styles: Use low-level cache for get_user_data_from_wp_global_styles

### PR DESCRIPTION
**Pull request by mmtr**
Backports changes from https://github.com/WordPress/wordpress-develop/pull/3517

See also https://github.com/WordPress/gutenberg/issues/44946 and https://core.trac.wordpress.org/ticket/56901

## Proposed changes
This change removes caching of global styles for logged in users, allowing "wp_global_styles" custom post type to be imported completely, regardless of any previously cached data. This change now relies on the lower-level native WP_Query cache invalidation methods for the global styles post type.

## Testing Instructions

- Use Twenty Twenty Two (TT2) theme
- Use WordPress 6.0.3
- Log into the admin area
- Go to Appearance > Editor
- Open the Global Styles UI
- Select "Browse Styles"
- Select any non-default variation
- Select "Save" to save the new global styles
- Go back to the Dashboard
- Go to Tools > Export
- Click on "Download Export File" using "All content"
- Using another local site, with WordPress 6.1 RC3 (where the issue https://github.com/WordPress/gutenberg/issues/44946 is still present) and Gutenberg running this branch
- Log into the admin area
- Go to Tools > Import
- Select "Install Now" for the "WordPress" importer
- Select "Run Importer"
- Select the XML file using "Choose File"
- Click on the "Upload file and import" button
- View site in the frontend
- Make sure the styles use the non-default variation you selected

### CODE REVIEW

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 1]
```diff
@@ -120,4 +120,64 @@ public static function get_user_data() {
...
+			'tax_query'   => array(
+				array(
+					'taxonomy' => 'wp_theme',
+					'field'    => 'name',
+					'terms'    => $theme->get_stylesheet(),
+				),
+			),
+		);
+
+		$recent_posts = wp_get_recent_posts( $args );
```
Comment (spacedmonkey) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php]: This method does not match core. `wp_get_recent_posts` was removed.

Comment (Mamaduka) [reply lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php]: Let's just fully copy the core version here - https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-theme-json-resolver.php#L402-L450

Comment (mmtr) [reply lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php]: Yup, just noted that! I updated it with the code that's available in WP 6.1.

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 2]
```diff
@@ -120,4 +120,72 @@ public static function get_user_data() {
...
+	public static function get_user_data_from_wp_global_styles( $theme, $create_post = false, $post_status_filter = array( 'publish' ) ) {
+		if ( ! $theme instanceof WP_Theme ) {
+			$theme = wp_get_theme();
+		}
+		$user_cpt         = array();
+		$post_type_filter = 'wp_global_styles';
+		$stylesheet       = $theme->get_stylesheet();
+		$args             = array(
+			'posts_per_page'      => 1,
+			'orderby'             => 'post_date',
```
Comment (oandregal) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php]:
```suggestion
			'orderby'             => 'date',
```

File: lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php [hunk 3]
```diff
@@ -120,4 +120,72 @@ public static function get_user_data() {
...
+		);
+
+		$global_style_query = new WP_Query();
+		$recent_posts       = $global_style_query->query( $args );
+		if ( count( $recent_posts ) === 1 ) {
+			$user_cpt = get_post( $recent_posts[0], ARRAY_A );
+		} elseif ( $create_post ) {
+			$cpt_post_id = wp_insert_post(
+				array(
+					'post_content' => '{"version": ' . WP_Theme_JSON::LATEST_SCHEMA . ', "isGlobalStylesUserThemeJSON": true }',
```
Comment (oandregal) [inline lib/compat/wordpress-6.1/class-wp-theme-json-resolver-6-1.php]:
```suggestion
					'post_content' => '{"version": ' . WP_Theme_JSON_Gutenberg::LATEST_SCHEMA . ', "isGlobalStylesUserThemeJSON": true }',
```

File: phpunit/class-wp-theme-json-resolver-test.php [hunk 4]
```diff
@@ -255,29 +309,53 @@ public function test_merges_child_theme_json_into_parent_theme_json() {
 		);
 	}
 
-	public function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries() {
+	function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries() {
```
Comment (mmtr) [inline phpunit/class-wp-theme-json-resolver-test.php]:
```suggestion
	public function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries() {
```

File: phpunit/class-wp-theme-json-resolver-test.php [hunk 5]
```diff
@@ -255,29 +309,53 @@ public function test_merges_child_theme_json_into_parent_theme_json() {
...
+			$this->assertSameSets( $user_cpt, $new_user_cpt, "User CPTs do not match on run {$i}." );
+		}
+		$query_count = count( $this->queries ) - $query_count;
+		$this->assertSame( 1, $query_count, 'Unexpected SQL queries detected for the wp_global_style post type after creation.' );
+	}
+
+	/**
+	 * @covers WP_Theme_JSON_Resolver::get_user_data_from_wp_global_styles
+	 */
+	function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries_for_logged_out_users() {
```
Comment (mmtr) [inline phpunit/class-wp-theme-json-resolver-test.php]:
```suggestion
	public function test_get_user_data_from_wp_global_styles_does_not_use_uncached_queries_for_logged_out_users() {
```

**Review by oandregal — APPROVED** [review]
Thanks for the PR!

This backports code that is in core, both in 6.1 and `trunk`. I've done some testing and it works fine: using style variations in both new sites (no wp_global_styles CPT) and old sites; changing themes; using WordPress 6.1 and 6.0 as base.

I was unable to reproduce the bug with import content in Gutenberg `trunk` (the issue that was raised in 6.1 RC3). This change is worth doing anyway.

**Pull request merged**